src/legendhpges/base.py

```diff
@@ -105,6 +110,65 @@ def _decode_polycone_coord(self) -> tuple[list[float], list[float]]:
         Must be overloaded by derived classes.
         """
 
+    def distance_to_surface(
+        self, coords: np.ndarray | list, surface_indices: list | None = None
+    ) -> np.ndarray:
+        """Compute the distance of a set of points to the nearest detector surface.
+
+        Parameters
+        ----------
+        coords
+            2D array of `(x,y,z)` coordinates for each point, first index corresponds to the point, second to the dimension `(x,y,z)`.
+        surface_indices
+            list of indices of surfaces to consider. If `None` (the default) all surfaces used.
+
+        Returns
+        -------
+        numpy array of the distance from each point to the nearest surface.
+
+        Note
+        ----
+        - Only implemented for solids based on :class:`G4GenericPolycone`
+        - Coordinates should be relative to the origin of the polycone.
+        """
+        # check type of the solid
+        if isinstance(self.solid, geant4.solid.GenericPolycone) is False:
+            msg = f"distance_to_surface is not implemented for {type(self.solid)} yet"
+            raise NotImplementedError(msg)
+
+        if not isinstance(coords, np.ndarray):
+            coords = np.array(coords)
+
+        if np.shape(coords)[1] != 3:
+            msg = "coords must be provided as a 2D array with x,y,z coordinates for each point."
+            raise ValueError(msg)
+
+        # convert x,y,z into r,z
+        rz_coords = utils.convert_coords(coords)
+
+        # get the coordinates
+        r, z = self._decode_polycone_coord()
```

Comment:
This is minor but you should try to get the cached polycone coordinates, instead of re-computing them.